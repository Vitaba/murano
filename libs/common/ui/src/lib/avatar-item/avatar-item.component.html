<pre>{{ data | json}} </pre>
<div [class]="styles.container">
  <ng-template #imageDefaultTemplate>
    <img
      [class]="styles.image"
      [src]="data.image.value"
      [alt]="data.image.alt"
    />
  </ng-template>
  <ng-container [ngTemplateOutlet]="imageExtraTemplate || imageDefaultTemplate">
  </ng-container>

  <div [class]="styles.dataContainer">
    <ng-template #nameDefaultTemplate>
      <p [class]="styles.name">{{ data.name }}</p>
    </ng-template>
    <ng-container [ngTemplateOutlet]="nameExtraTemplate || nameDefaultTemplate">
    </ng-container>

    <ng-template #dateDefaultTemplate>
      <p [class]="styles.date">{{ data.date | date : config.dateFormat}}</p>
    </ng-template>
    <ng-container [ngTemplateOutlet]="dateExtraTemplate || dateDefaultTemplate">
    </ng-container>
    
  </div>
</div>
