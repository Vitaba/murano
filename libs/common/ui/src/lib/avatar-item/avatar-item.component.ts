import { ChangeDetectionStrategy, Component, ContentChild, Input, TemplateRef } from '@angular/core';
interface AvatarConfig {
  dateFormat?: string;
}

interface AvatarData {
  name: string;
  image: {
    value: string;
    alt: string;
  };
  date?: string;
  dateFormat?: string;
}

interface AvatarStyles {
  container: string;
  image: string;
  dataContainer: string;
  name: string;
  date: string;
}

@Component({
  changeDetection: ChangeDetectionStrategy.OnPush,
  selector: 'vitaba-avatar-item',
  styleUrls: ['./avatar-item.component.scss'],
  templateUrl: './avatar-item.component.html',
})

export class AvatarItemComponent {
  @Input() public config: AvatarConfig = {
    dateFormat: 'mediumDate',
  };
  @Input() public styles: AvatarStyles = {
    container: 'flex items-center',
    dataContainer: 'text-sm',
    date: 'text-gray-600',
    image: 'w-10 h-10 rounded-full mr-4',
    name: 'text-gray-900 leading-none',
  };
  @Input() public data: AvatarData = {
    date: '2019-11-26T15:03:29.435Z',
    image: {
      alt: 'Avatar de Sebastian',
      value: 'https://avatars3.githubusercontent.com/u/17608169?s=460&v=4',
    },
    name: 'Sebastian'};

  @ContentChild('imageExtraTemplate', { static: false })
  // tslint:disable-next-line: no-any
  public imageExtraTemplate!: TemplateRef<any>;

  @ContentChild('nameExtraTemplate', { static: false })
  // tslint:disable-next-line: no-any
  public nameExtraTemplate!: TemplateRef<any>;

  @ContentChild('dateExtraTemplate', { static: false })
  // tslint:disable-next-line: no-any
  public dateExtraTemplate!: TemplateRef<any>;
}
