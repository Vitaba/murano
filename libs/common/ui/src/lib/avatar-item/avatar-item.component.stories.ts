import { object } from '@storybook/addon-knobs';
import { AvatarItemComponent } from './avatar-item.component';
// tslint:disable-next-line:no-default-export
export default {
  title: 'AvatarItemComponent',
};

const data = {
  date: '2019-11-26T15:03:29.435Z',
  image: {
    alt: 'Avatar de Jhon Doe',
    value: 'https://avatars1.githubusercontent.com/u/7867954?s=400&v=4',
  },
  name: 'Jhon Doe',
};

const styles = {
  container: 'flex items-center',
  dataContainer: 'text-sm',
  date: 'text-gray-600',
  image: 'w-10 h-10 rounded-full mr-4',
  name: 'text-gray-900 leading-none',
};

const config = {
  dateFormat: 'mediumDate',
};

export const basic = () => ({
  moduleMetadata: {
    declarations: [AvatarItemComponent],
    imports: [],
  },
  props: {
    config: object('styles', { ...config }),
    data: object('data', data),
    styles: object('styles', { ...styles }),
  },
  template: `<vitaba-avatar-item [data]="data"> </vitaba-avatar-item>`,
});

export const imageProjection = () => ({
  moduleMetadata: {
    declarations: [AvatarItemComponent],
    imports: [],
  },
  props: {
    config: object('styles', { ...config }),
    data: object('data', { ...data, image: undefined }),
    styles: object('styles', { ...styles }),
  },
  template: `<vitaba-avatar-item [data]="data">
  <ng-template #imageExtraTemplate> <b> Image Extra Template </b> </ng-template>
  </vitaba-avatar-item>`,
});

export const nameProjection = () => ({
  moduleMetadata: {
    declarations: [AvatarItemComponent],
    imports: [],
  },
  props: {
    config: object('styles', { ...config }),
    data: object('data', { ...data, name: undefined }),
    styles: object('styles', { ...styles }),
  },
  template: `<vitaba-avatar-item [data]="data">
  <ng-template #nameExtraTemplate> <b> Name Extra Template </b> </ng-template>
  </vitaba-avatar-item>`,
});

export const dateProjection = () => ({
  moduleMetadata: {
    declarations: [AvatarItemComponent],
    imports: [],
  },
  props: {
    config: object('styles', { ...config }),
    data: object('data', { ...data, date: undefined }),
    styles: object('styles', { ...styles }),
  },
  template: `<vitaba-avatar-item [data]="data">
  <ng-template #dateExtraTemplate> <b>Date content-projection</b> </ng-template>
  </vitaba-avatar-item>`,
});
