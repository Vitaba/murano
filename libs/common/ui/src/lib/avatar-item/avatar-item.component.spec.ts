
// tslint:disable:component-max-inline-declarations ter-max-len
import { ChangeDetectionStrategy, Component, DebugElement } from '@angular/core';
import { async, ComponentFixture, TestBed } from '@angular/core/testing';
import { AvatarItemComponent } from './avatar-item.component';

@Component({
  changeDetection: ChangeDetectionStrategy.OnPush,
  selector: 'vitaba-avatar-item-sample-implementation',
  template: `
  <ng-container [ngSwitch]="mode">
    <vitaba-avatar-item *ngSwitchDefault [data]="avatarData" [styles]="avatarStyles" [config]="avatarConfig">
    </vitaba-avatar-item>

    <vitaba-avatar-item *ngSwitchCase="'date'" [data]="avatarData" [styles]="avatarStyles" [config]="avatarConfig">
      <ng-template #dateExtraTemplate>
        <p>Date content-projection</p>
      </ng-template>
    </vitaba-avatar-item>

    <vitaba-avatar-item *ngSwitchCase="'name'" [data]="avatarData" [styles]="avatarStyles" [config]="avatarConfig">
      <ng-template #nameExtraTemplate>
        <p>Name content-projection</p>
      </ng-template>
    </vitaba-avatar-item>

    <vitaba-avatar-item *ngSwitchCase="'image'" [data]="avatarData" [styles]="avatarStyles" [config]="avatarConfig">
      <ng-template #imageExtraTemplate>
        <p>Image content-projection</p>
      </ng-template>
    </vitaba-avatar-item>
  </ng-container>
  `,
})
class TestHostComponent {
  public avatarData = {
    date: '2019-11-26T15:03:29.435Z',
    image: {
      alt: 'John Doe Avatar',
      value: 'https://avatars3.githubusercontent.com/u/17608169?s=460&v=4',
    },
    name: 'John Doe',
  };
  public avatarConfig = {
    dateFormat: 'mediumDate',
  };
  public avatarStyles = {
    container: 'flex items-center',
    dataContainer: 'text-sm',
    date: 'text-gray-600',
    image: 'w-10 h-10 rounded-full mr-4',
    name: 'text-gray-900 leading-none',
  };
  public mode = 'default';
}

describe('AvatarItemComponent', () => {
  let component: TestHostComponent;
  let fixture: ComponentFixture<TestHostComponent>;
  let debugElement: DebugElement;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      declarations: [TestHostComponent, AvatarItemComponent],
    })
    .compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(TestHostComponent);
    component = fixture.componentInstance;
    debugElement = fixture.debugElement;
  });

  it('should create', () => {
    fixture.detectChanges();
    expect(component).toBeTruthy();
  });

  it('should be default', () => {
    fixture.changeDetectorRef.detectChanges();
    expect(fixture).toMatchSnapshot();
  });

  it('should be imageExtraTemplate', () => {
    component.mode = 'image';
    fixture.changeDetectorRef.detectChanges();
    expect(fixture).toMatchSnapshot();
  });

  it('should be nameExtraTemplate', () => {
    component.mode = 'name';
    fixture.changeDetectorRef.detectChanges();
    expect(fixture).toMatchSnapshot();
  });

  it('should be dateExtraTemplate', () => {
    component.mode = 'date';
    fixture.changeDetectorRef.detectChanges();
    expect(fixture).toMatchSnapshot();
  });

  it('should throw errors when data is not in right format', () => {
    component.mode = 'default';
    component.avatarData = {
      date: 1,
      image: {
        alt: 1,
        value: 1,
      },
      name: 1,
    } as any;
    fixture.changeDetectorRef.detectChanges();
    expect(fixture).toMatchSnapshot();
  });
});
