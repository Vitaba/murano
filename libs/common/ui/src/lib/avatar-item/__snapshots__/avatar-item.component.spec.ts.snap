// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AvatarItemComponent should be dateExtraTemplate 1`] = `
<vitaba-avatar-item-sample-implementation
  avatarConfig={[Function Object]}
  avatarData={[Function Object]}
  avatarStyles={[Function Object]}
  mode={[Function String]}
>
  <vitaba-avatar-item>
    
    
    <div
      class="flex items-center"
    >
      
      
      <img
        alt="John Doe Avatar"
        class="w-10 h-10 rounded-full mr-4"
        src="https://avatars3.githubusercontent.com/u/17608169?s=460&v=4"
      />
      <div
        class="text-sm"
      >
        
        
        <p
          class="text-gray-900 leading-none"
        >
          John Doe
        </p>
        
        
        <p>
          Date content-projection
        </p>
      </div>
    </div>
    
  </vitaba-avatar-item>
</vitaba-avatar-item-sample-implementation>
`;

exports[`AvatarItemComponent should be default 1`] = `
<vitaba-avatar-item-sample-implementation
  avatarConfig={[Function Object]}
  avatarData={[Function Object]}
  avatarStyles={[Function Object]}
  mode={[Function String]}
>
  <vitaba-avatar-item>
    
    
    <div
      class="flex items-center"
    >
      
      
      <img
        alt="John Doe Avatar"
        class="w-10 h-10 rounded-full mr-4"
        src="https://avatars3.githubusercontent.com/u/17608169?s=460&v=4"
      />
      <div
        class="text-sm"
      >
        
        
        <p
          class="text-gray-900 leading-none"
        >
          John Doe
        </p>
        
        
        <p
          class="text-gray-600"
        >
          Nov 26, 2019
        </p>
      </div>
    </div>
    
  </vitaba-avatar-item>
</vitaba-avatar-item-sample-implementation>
`;

exports[`AvatarItemComponent should be imageExtraTemplate 1`] = `
<vitaba-avatar-item-sample-implementation
  avatarConfig={[Function Object]}
  avatarData={[Function Object]}
  avatarStyles={[Function Object]}
  mode={[Function String]}
>
  <vitaba-avatar-item>
    
    
    <div
      class="flex items-center"
    >
      
      
      <p>
        Image content-projection
      </p>
      <div
        class="text-sm"
      >
        
        
        <p
          class="text-gray-900 leading-none"
        >
          John Doe
        </p>
        
        
        <p
          class="text-gray-600"
        >
          Nov 26, 2019
        </p>
      </div>
    </div>
    
  </vitaba-avatar-item>
</vitaba-avatar-item-sample-implementation>
`;

exports[`AvatarItemComponent should be nameExtraTemplate 1`] = `
<vitaba-avatar-item-sample-implementation
  avatarConfig={[Function Object]}
  avatarData={[Function Object]}
  avatarStyles={[Function Object]}
  mode={[Function String]}
>
  <vitaba-avatar-item>
    
    
    <div
      class="flex items-center"
    >
      
      
      <img
        alt="John Doe Avatar"
        class="w-10 h-10 rounded-full mr-4"
        src="https://avatars3.githubusercontent.com/u/17608169?s=460&v=4"
      />
      <div
        class="text-sm"
      >
        
        
        <p>
          Name content-projection
        </p>
        
        
        <p
          class="text-gray-600"
        >
          Nov 26, 2019
        </p>
      </div>
    </div>
    
  </vitaba-avatar-item>
</vitaba-avatar-item-sample-implementation>
`;

exports[`AvatarItemComponent should throw errors when data is not in right format 1`] = `
<vitaba-avatar-item-sample-implementation
  avatarConfig={[Function Object]}
  avatarData={[Function Object]}
  avatarStyles={[Function Object]}
  mode={[Function String]}
>
  <vitaba-avatar-item>
    
     {
    "valid": false,
    "guard": "isAvatarData",
    "value": {
      "image": 1,
      "name": 1
    },
    "message": "Is not a avatar data",
    "errors": [
      {
        "valid": false,
        "guard": "hasProperties",
        "value": {
          "image": 1,
          "name": 1
        },
        "errors": [
          {
            "value": {
              "name": "image",
              "type": "object"
            },
            "guard": "Object Validation",
            "message": "Is not an object",
            "valid": false
          },
          {
            "value": {
              "name": "name",
              "type": "string"
            },
            "guard": "String Validation",
            "message": "Is not a string",
            "valid": false
          },
          {
            "value": {
              "name": "date",
              "type": "string"
            },
            "guard": "String Validation",
            "message": "Is not a string",
            "valid": false
          }
        ]
      }
    ]
  }
  
    
  </vitaba-avatar-item>
</vitaba-avatar-item-sample-implementation>
`;
